as-ps/as_ps/main.py

```diff
@@ -100,64 +110,99 @@ def write(self, reason, value):
-        # Return dict indexed with reason
-        data, updated = bbb.read(device_name)
-        if not updated:
-            return
-        for reason, new_value in data.items():
-            if self._check_value_changed(reason, new_value):
-                if new_value is None:
-                    continue
-                self.driver.setParam(reason, new_value)
-            self.driver.setParamStatus(
-                reason, _Alarm.NO_ALARM, _Severity.NO_ALARM)
+    def _write_operation(self, bbb, pvname, value):
+        t0 = _time.time()
+        bbb.write(pvname.device_name, pvname.propty, value)
+        # NOTE: This scan_device might be redundent
+        self._scan_device(bbb, pvname.device_name, force_update=True)
+        t1 = _time.time()
+        _log.info("[{:.2s}] - {:.32s} = {:.50s}".format(
+            'T ', pvname, 'write operation took {:.3f} ms'.format((t1-t0)*1000)))
 
-    def _update_ioc_database_disconnected(self, bbb, device_name):
+    def _scan_bbb(self, bbb):
+        for device_name in bbb.psnames:
+            self._scan_device(bbb, device_name)
+
+    def _scan_device(self, bbb, device_name, force_update=False):
+        bbb_connected = bbb.check_connected(device_name)
+        self._update_ioc_database(bbb, device_name, bbb_connected, force_update)
+
+    def _update_ioc_database(self, bbb, device_name,
+                             bbb_connected=True,
+                             force_update=False):
         # Return dict indexed with reason
-        data, updated = bbb.read(device_name)
+        data, updated = bbb.read(device_name, force_update=force_update)
+        # data, updated = self._bbb_read(bbb, device_name,
+        #                                force_update=force_update)
         if not updated:
             return
+
         for reason, new_value in data.items():
+            if self._prucqueue and App._setpoint_regexp.match(reason):
+                # While there are pending write operations in the queue we
+                # cannot update setpoint variables or we will spoil the
+                # accepted value in the write method.
+                continue
             if self._check_value_changed(reason, new_value):
                 if new_value is None:
                     continue
                 self.driver.setParam(reason, new_value)
-            self.driver.setParamStatus(
-                reason, _Alarm.TIMEOUT_ALARM, _Severity.INVALID_ALARM)
-
-    def _scan_bbb(self, bbb):
-        for device_name in bbb.psnames:
-            if bbb.check_connected(device_name):
-                self._update_ioc_database(bbb, device_name)
+            if bbb_connected:
+                self.driver.setParamStatus(
+                    reason, _Alarm.NO_ALARM, _Severity.NO_ALARM)
             else:
-                self._update_ioc_database_disconnected(bbb, device_name)
+                self.driver.setParamStatus(
+                    reason, _Alarm.TIMEOUT_ALARM, _Severity.INVALID_ALARM)
+
```

Comment:
I think you have to call `self.driver.updatePV` in the end of this method, dont' you?

Reply:
Not strictly necessary since all invoking functions of `_scan_bbb` do update the epics PVs database right after invoking the function. But maybe it is better to do it inside `_scan_bbb` to simplify its interpretation. I have done it in a new commit and renamed the function to `scan_bbb_updatePVs`, to make it clearer.

Reply:
Humm... I still think it is better to call here, because the name of the method is `_update_ioc_database` and here is where you write the new values to the database. Besides, if you do it here, you'll only need to call `updatePV` and not `updatePVs`, which is slower...

Reply:
I will unresolve this one for now, but if you still think it is better to do the way it is now, feel free to resolve it.

Reply:
@fernandohds564 , I am confused. What exactly do you mean "I still think it is better to call here", inside `_scan_device` or  inside `_update_ioc_database`? in boths cases you would be calling `updatePVs` twice from the driving `scan_bbb` method. 
Fernando, I am resolving this one. :grimacing:

Reply:
What I mean by "here" is where I've placed the comment: inside the method `_update_ioc_database`.
You wouldn't have to call `updatePVs` at all if you called `updatePV`inside the above mentioned method...

Reply:
Ok. I thought `updatePVs` was more efficient (implemented in C) than many calls to `updatePV`. Implementing your suggestion...